config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java

```diff
@@ -0,0 +1,95 @@
+package com.typesafe.config.impl;
+
+import com.typesafe.config.ConfigNode;
+
+import java.util.*;
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
+    final private LinkedHashMap<Path, Integer> map = new LinkedHashMap<>();
+    final ArrayList<Integer> keyValueIndexes;
+
+    ConfigNodeComplexValue(Collection<ConfigNode> children) {
+        this.children = new ArrayList(children);
+        keyValueIndexes = new ArrayList<Integer>();
+
+        // Construct the list of indexes of Key-Value nodes. Do this
+        // in reverse order, since all but the final duplicate will be removed.
+        for (int i = this.children.size() - 1; i >= 0; i--) {
+            ConfigNode currNode = this.children.get(i);
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
+    }
+
+    public ArrayList<ConfigNode> children() {
+        return children;
+    }
+
+    public String render() {
+        StringBuilder renderedText = new StringBuilder();
+        for (ConfigNode child : children) {
+            renderedText.append(child.render());
+        }
+        return renderedText.toString();
```

Comment:
oh, but if render() always just means "render all the tokens", this can be even simpler: have a final render method that calls `tokens` and the subtypes only have to implement `tokens()`